astropy/units/decorators.py

```diff
@@ -115,26 +117,51 @@ def wrapper(*func_args, **func_kwargs):
                 # If the target unit is empty, then no unit was specified so we
                 # move past it
                 if target_unit is not funcsigs.Parameter.empty:
-                    try:
-                        equivalent = arg.unit.is_equivalent(target_unit,
-                                                  equivalencies=self.equivalencies)
-
-                        if not equivalent:
-                            raise UnitsError("Argument '{0}' to function '{1}'"
-                                             " must be in units convertible to"
-                                             " '{2}'.".format(param.name,
+
+                    if isiterable(target_unit):
+                        target_units = target_unit
+                        no_equivalent_msg = ("Argument '{0}' to function '{1}' must be in "
+                                             "units convertible to one of: {2}."
+                                             .format(param.name,
+                                                     wrapped_function.__name__,
+                                                     [uu.to_string() for uu in target_units
+                                                      if uu is not None]))
+                    else:
+                        target_units = [target_unit]
+                        no_equivalent_msg = ("Argument '{0}' to function '{1}' must be in "
+                                             "units convertible to '{2}'."
+                                             .format(param.name,
                                                      wrapped_function.__name__,
                                                      target_unit.to_string()))
 
-                    # Either there is no .unit or no .is_equivalent
-                    except AttributeError:
-                        if hasattr(arg, "unit"):
-                            error_msg = "a 'unit' attribute without an 'is_equivalent' method"
-                        else:
-                            error_msg = "no 'unit' attribute"
-                        raise TypeError("Argument '{0}' to function '{1}' has {2}. "
-                              "You may want to pass in an astropy Quantity instead."
-                                 .format(param.name, wrapped_function.__name__, error_msg))
+                    # short circuit if optional argument
+                    if arg is None:
```

Comment:
I would say that we should allow `None` only if it's the default value of a keyword argument.